doc/developer/20250723_configurable_replica_sizes.md

```diff
@@ -0,0 +1,109 @@
+# Configurable Cluster Replica Sizes
+
+## The Problem
+
+We want users to be able to safely configure cluster replica sizes on runtime for self managed.
+
+> At the moment, we pass the cluster replica size map as a command line argument to environmentd, and this is the only way to specify cluster replica sizes and their metadata. A change to the cluster replica size map requires a restart of environmentd, and patching the Kubernetes stateful set, or whatever deployment mechanism the user has in place. This is a tedious, dangerous, and inefficient process.
+>
+
+Currently, cluster sizes are created on bootstrap via CLI arguments, saved in the catalog as `cluster_replica_sizes`, then saved to our builtin tables. On creation of a cluster replica, we save it as a `ReplicaAllocation` object [[code pointer](https://github.com/MaterializeInc/materialize/blob/d37c5be00a29b8b4d5e341f3efac15f9f01a8942/src/adapter/src/catalog/state.rs#L2290)]
+
+## Success Criteria
+
+- Allow a self managed user to configure cluster replica sizes with a schema similar to this code snippet: [https://materialize.com/docs/self-managed/v25.2/sql/appendix-cluster-sizes/#custom-cluster-sizes](https://materialize.com/docs/self-managed/v25.2/sql/appendix-cluster-sizes/#custom-cluster-sizes)
+- Allow a user to validate the custom sizes by interacting with the database itself
+- Fix potential breaking changes in all clients (i.e. Console)
+- Update Terraform providers to use this new functionality
+
+## Out of Scope
+
+- Allow configurable sizes in Cloud
```

Comment:
What's the reason for not wanting to consider cloud? For the rollout of swap, being able to specify the replica sizes in dyncfgs would be a huge boon. We could set up LD to perform a slow rollout of swap, based on the risk segment and the Mz version.

Reply:
> Only a single self managed potential customer has asked for it

I'd view asks from the team help our internal management of MZ cloud at about level as "customer ask" in this case. If the ways customers and MZ would use these are very different then we may need to separate, but LD vs file dyncfg are close enough. If customers would really want `create replica size` that maybe takes on a life of its own.

Reply:
The latest thinking is that we might introduce new replica sizes, instead of swapping out the spilling mechanism for `cc` sizes, in which case this might not be relevant for the swap rollout after all. It still seems useful to be able to configure replica sizes through LD, to make one-off adjustments without having to manually edit statefulsets and hope the changes stick around for long enough.